cardano-db-sync/src/Cardano/DbSync/Plugin/Epoch.hs

```diff
@@ -92,36 +100,80 @@ epochPluginInsertBlock trce blkTip = do
-                 updateEpochNum epochNum trce
-            else pure $ Right ()
-
-      | epochNum > 0 && mLatestCachedEpoch == Nothing ->
-          updateEpochNum 0 trce
-      | epochNum >= lastCachedEpoch + 2 ->
-          updateEpochNum (lastCachedEpoch + 1) trce
-      | epochNum == chainTipEpoch && lastCachedEpoch < chainTipEpoch ->
-          updateEpochNum (lastCachedEpoch + 1) trce
-      | epochNum > chainTipEpoch ->
-          -- Must just have started a new epoch, so call this which will
-          -- update chainTipEpoch.
-          updateEpochNum epochNum trce
-      | otherwise ->
-          pure $ Right ()
+
+  let blockDiff = unBlockNo tipNo - blockNo
+
+  let updateEpochAction :: UpdateEpochCmd
+      updateEpochAction = updateEpochNumLogic epochNum mLatestCachedEpoch chainTipEpoch blockDiff
+
+  -- Convert commands into actions.
+  case updateEpochAction of
+    UpdateEpoch epochNum' -> do
+        -- For debugging! TO BE REMOVED!
+        liftIO . logInfo trce $ "UpdateEpoch!"
+        liftIO . logInfo trce $ "    epochNum = " <> (show epochNum :: Text)
+        liftIO . logInfo trce $ "    mLatestCachedEpoch = " <> (show mLatestCachedEpoch :: Text)
+        liftIO . logInfo trce $ "    chainTipEpoch = " <> (show chainTipEpoch :: Text)
+        liftIO . logInfo trce $ "    blockDiff = " <> (show blockDiff :: Text)
+        updateEpochNumDefault epochNum' trce
+    DoNotUpdate -> pure $ Right ()
+
+-- |Commands for separating the intructions from it's effect.
+data UpdateEpochCmd
+  = UpdateEpoch Word64
+  | DoNotUpdate
+  deriving (Eq, Show)
+
+-- |Update logic for the epochs.
+updateEpochNumLogic
+  :: Word64
+  -> Maybe Word64
+  -> Word64
+  -> Word64
+  -> UpdateEpochCmd
+updateEpochNumLogic epochNum mLatestCachedEpoch chainTipEpoch blockDiff = do
+
+      let lastCachedEpoch = fromMaybe 0 mLatestCachedEpoch
+
+      if  | epochNum == chainTipEpoch && lastCachedEpoch == chainTipEpoch ->
+              -- Following the chain very closely.
+              if blockDiff < 15
+                then UpdateEpoch epochNum
+                else DoNotUpdate
+
+          | epochNum > 0 && mLatestCachedEpoch == Nothing ->
+              -- There is no cache of the epoch.
+              UpdateEpoch 0
+
+          | epochNum >= lastCachedEpoch + 2 ->
+              -- The diff for the cached and current epoch is >= 2.
+              UpdateEpoch (lastCachedEpoch + 1)
+
+          | epochNum == chainTipEpoch && lastCachedEpoch < chainTipEpoch ->
+              UpdateEpoch (lastCachedEpoch + 1)
+
+          | epochNum > chainTipEpoch ->
+              -- Must just have started a new epoch, so call this which will
+              -- update chainTipEpoch.
+              -- This can hit the second condition and the third one.
+              -- It's very weird that this is here.
+              UpdateEpoch epochNum
+
+          | otherwise ->
+              -- I really don't see what could hit this one.
```

Comment:
When the chain is syncing (as opposed to closely following) we only want to update the required row in the Epoch table once instead of on every new block. This the case in this otherwise cause.